File: Assets/Scripts/MeterFireExtinguisher.cs

// Hannes Helmholz
//
// 

using UnityEngine;
using System.Collections;
using Holoville.HOTween.Core;
using Holoville.HOTween.Plugins;
using Holoville.HOTween.Plugins.Core;
using Holoville.HOTween;

public class MeterFireExtinguisher : MeterAbstract
{

    // =============================================================================
    // MEMBERS ---------------------------------------------------------------------
    [SerializeField]
    private float ExtinguishAnimationLength = 6;
    private LayerSetterElement Nose;
    // =============================================================================
 
 
 
    // =============================================================================
    // METHODS UNITY ---------------------------------------------------------------
 
    void Awake ()
    {
        base.Awake ();
     
        Nose = gameObject.GetComponentInChildren<LayerSetterElement> ();
     
        Audio.clip = SoundSuccess;
    }
    // =============================================================================
 
 
 
    // =============================================================================
    // METHODS ---------------------------------------------------------------------

    public IEnumerator Extinguish ( Transform port )
    {
        if ( !IsAvailable () )
        {
            return false;
        }
     
        float measureAnimationPartLength = ExtinguishAnimationLength/4.0f;
     
        SetPlayerMotion ( false );
     
        TweenParms parms1 = new TweenParms ();

        parms1.Prop ( "position", new PlugVector3 ( port.position ) );
        parms1.Prop ( "rotation", new PlugQuaternion ( port.rotation ) );
        //parms1.Prop("localScale", Plugins.Core.PlugVector3(port.localScale));
        parms1.Ease ( EaseType.EaseInOutSine );
        Tweener tweener1 = HOTween.To ( Trans, measureAnimationPartLength, parms1 );
        tweener1.autoKillOnComplete = false;

        //dan
        //TODO
        //yield WaitForSeconds(measureAnimationPartLength);
        yield return new WaitForSeconds(measureAnimationPartLength);

        LayerSetter.SetAllLayerRecursively ( Nose.gameObject, Nose.OtherLayer );

        TweenParms parms2 = new TweenParms ();
        parms2.Prop ( "position",  new PlugVector3 ( Trans.forward*0.18f, true ) );
        parms2.Ease ( EaseType.EaseInOutSine );
        Tweener tweener2 = HOTween.To ( Trans, measureAnimationPartLength, parms2 );
        tweener2.autoKillOnComplete = false;

        //dan
        //TODO
        //yield WaitForSeconds(measureAnimationPartLength);
        yield return new WaitForSeconds(measureAnimationPartLength);

        //dan
        //TODO
        //yield base.MeasureValues(3); // play sound
        yield return base.MeasureValues(3); // play sound


        //dan
        //TODO
        //yield WaitForSeconds(0.5f);
        yield return new WaitForSeconds(0.5f);
     
        HOTween.PlayBackwards ( tweener2 );

        //dan
        //TODO
        //yield WaitForSeconds(measureAnimationPartLength);
        yield return new WaitForSeconds(measureAnimationPartLength);

        LayerSetter.SetAllLayerRecursively ( Nose.gameObject, Nose.OldLayer );
     
        HOTween.PlayBackwards ( tweener1 );

        //dan
        //TODO
        //yield WaitForSeconds(measureAnimationPartLength);
        yield return new WaitForSeconds(measureAnimationPartLength);


        HOTween.Kill ( tweener1 );
        HOTween.Kill ( tweener2 );
        SetPlayerMotion ( true );
    }
    // =============================================================================
}

File: ISS Emergency/Assets/Scripts/MeterFireExtinguisher.js
#pragma strict

// Hannes Helmholz
//
// 


private class MeterFireExtinguisher extends MeterAbstract {

	// =============================================================================
	// MEMBERS ---------------------------------------------------------------------
	@SerializeField private var ExtinguishAnimationLength:float = 6;

	private var Nose:LayerSetterElement;
	// =============================================================================
	
	
	
	// =============================================================================
	// METHODS UNITY ---------------------------------------------------------------
	
	function Awake() {
		super.Awake();
		
		Nose = gameObject.GetComponentInChildren.<LayerSetterElement>();
		
		Audio.clip = SoundSuccess;
	}
	// =============================================================================
	
	
	
	// =============================================================================
	// METHODS ---------------------------------------------------------------------

	public function Extinguish(port:Transform) {
		if (!IsAvailable())
			return;
		
		var measureAnimationPartLength:float = ExtinguishAnimationLength / 4.0;
		
		SetPlayerMotion(false);
		
		var parms1:TweenParms = TweenParms();
		parms1.Prop("position", Plugins.Core.PlugVector3(port.position));
		parms1.Prop("rotation", Plugins.PlugQuaternion(port.rotation));
		//parms1.Prop("localScale", Plugins.Core.PlugVector3(port.localScale));
		parms1.Ease(EaseType.EaseInOutSine);
		var tweener1:Tweener = HOTween.To(Trans, measureAnimationPartLength, parms1);
		tweener1.autoKillOnComplete = false;
		yield WaitForSeconds(measureAnimationPartLength);
		
		LayerSetter.SetAllLayerRecursively(Nose.gameObject, Nose.OtherLayer);
		
		var parms2:TweenParms = TweenParms();
		parms2.Prop("position", Plugins.Core.PlugVector3(Trans.forward * 0.18, true));
		parms2.Ease(EaseType.EaseInOutSine);
		var tweener2:Tweener = HOTween.To(Trans, measureAnimationPartLength, parms2);
		tweener2.autoKillOnComplete = false;
		yield WaitForSeconds(measureAnimationPartLength);
		
		yield super.MeasureValues(3); // play sound
		yield WaitForSeconds(0.5);
		
		HOTween.PlayBackwards(tweener2);
		yield WaitForSeconds(measureAnimationPartLength);
		
		LayerSetter.SetAllLayerRecursively(Nose.gameObject, Nose.OldLayer);
		
		HOTween.PlayBackwards(tweener1);
		yield WaitForSeconds(measureAnimationPartLength);
		
		HOTween.Kill(tweener1);
		HOTween.Kill(tweener2);
		SetPlayerMotion(true);
	}
	// =============================================================================
}